File: security_cam/test/gray_frame_assembler_test.dart
import 'dart:typed_data';

import 'package:flutter_test/flutter_test.dart';
import 'package:security_cam/sensors/gray_frame_assembler.dart';

void main() {
  test('emits a frame from a single exact-size chunk', () {
    final a = GrayFrameAssembler(2, 2);
    final frames = a.add(Uint8List.fromList([1, 2, 3, 4]));
    expect(frames.length, 1);
    expect(frames.single.gray, [1, 2, 3, 4]);
    expect(a.buffered, 0);
  });

  test('emits frames from arbitrary byte-split chunks', () {
    final a = GrayFrameAssembler(2, 2);
    final input = List<int>.generate(12, (i) => i); // 3 frames of 4 bytes
    final chunks = <Uint8List>[];
    for (var i = 0; i < input.length; i++) {
      chunks.add(Uint8List.fromList([input[i]]));
    }
    final frames = chunks.expand(a.add).toList();
    expect(frames.length, 3);
    expect(frames[0].gray, [0, 1, 2, 3]);
    expect(frames[2].gray, [8, 9, 10, 11]);
    expect(a.buffered, 0);
  });

  test('one chunk containing multiple frames emits all of them', () {
    final a = GrayFrameAssembler(2, 1);
    final frames = a.add(Uint8List.fromList([1, 2, 3, 4, 5, 6]));
    expect(frames.length, 3);
    expect(frames.map((f) => f.gray).toList(), [
      [1, 2],
      [3, 4],
      [5, 6],
    ]);
  });

  test('partial chunk carries remainder across calls', () {
    final a = GrayFrameAssembler(2, 2);
    expect(a.add(Uint8List.fromList([1, 2, 3])), isEmpty);
    expect(a.buffered, 3);
    final frames = a.add(Uint8List.fromList([4, 5]));
    expect(frames.length, 1);
    expect(frames.single.gray, [1, 2, 3, 4]);
    expect(a.buffered, 1);
    expect(a.add(Uint8List.fromList([6])), isEmpty);
    expect(a.buffered, 2);
    final rest = a.add(Uint8List.fromList([7, 8]));
    expect(rest.length, 1);
    expect(rest.single.gray, [5, 6, 7, 8]);
    expect(a.buffered, 0);
  });
}

File: level1/android/app/src/test/kotlin/io/securitycam/level1/sensors/GrayFrameAssemblerTest.kt
package io.securitycam.level1.sensors

import org.junit.Assert.assertEquals
import org.junit.Test

/** Port of `test/gray_frame_assembler_test.dart`. */
class GrayFrameAssemblerTest {

    @Test
    fun emitsAFrameFromASingleExactSizeChunk() {
        val a = GrayFrameAssembler(2, 2)
        val frames = a.add(byteArrayOf(1, 2, 3, 4))
        assertEquals(1, frames.size)
        assertEquals(listOf(1, 2, 3, 4), frames.single().gray.map { it.toInt() and 0xFF })
        assertEquals(0, a.buffered)
    }

    @Test
    fun emitsFramesFromArbitraryByteSplitChunks() {
        val a = GrayFrameAssembler(2, 2)
        val input = Array(12) { it.toByte() } // 3 frames of 4 bytes
        val frames = input.flatMap { a.add(byteArrayOf(it)) }
        assertEquals(3, frames.size)
        assertEquals(listOf(0, 1, 2, 3), frames[0].gray.map { it.toInt() and 0xFF })
        assertEquals(listOf(8, 9, 10, 11), frames[2].gray.map { it.toInt() and 0xFF })
        assertEquals(0, a.buffered)
    }

    @Test
    fun oneChunkContainingMultipleFramesEmitsAllOfThem() {
        val a = GrayFrameAssembler(2, 1)
        val frames = a.add(byteArrayOf(1, 2, 3, 4, 5, 6))
        assertEquals(3, frames.size)
        assertEquals(
            listOf(listOf(1, 2), listOf(3, 4), listOf(5, 6)),
            frames.map { f -> f.gray.map { it.toInt() and 0xFF } },
        )
    }

    @Test
    fun partialChunkCarriesRemainderAcrossCalls() {
        val a = GrayFrameAssembler(2, 2)
        assertEquals(emptyList<Any>(), a.add(byteArrayOf(1, 2, 3)))
        assertEquals(3, a.buffered)
        val frames = a.add(byteArrayOf(4, 5))
        assertEquals(1, frames.size)
        assertEquals(listOf(1, 2, 3, 4), frames.single().gray.map { it.toInt() and 0xFF })
        assertEquals(1, a.buffered)
        assertEquals(emptyList<Any>(), a.add(byteArrayOf(6)))
        assertEquals(2, a.buffered)
        val rest = a.add(byteArrayOf(7, 8))
        assertEquals(1, rest.size)
        assertEquals(listOf(5, 6, 7, 8), rest.single().gray.map { it.toInt() and 0xFF })
        assertEquals(0, a.buffered)
    }
}